pkg/node/node.go

```diff
@@ -101,105 +103,261 @@ func createAddressFilter(
 	}
 }
 
-// GenerateSignerForKeep generates a new threshold signer with ECDSA key pair. The
-// public key is a public key of the signing group. It publishes the public key
-// to the keep. It uses keep address as unique signing group identifier.
+// GenerateSignerForKeep generates a new threshold signer with ECDSA key pair
+// and submits the public key to the on-chain keep.
+//
+// The attempt for generating signer is retried on failure until the provided
+// context is done.
 func (n *Node) GenerateSignerForKeep(
 	ctx context.Context,
 	operatorPublicKey *operator.PublicKey,
 	keepAddress common.Address,
-	keepMembersIDs []tss.MemberID,
+	members []common.Address,
 ) (*tss.ThresholdSigner, error) {
 	memberID := tss.MemberIDFromPublicKey(operatorPublicKey)
 
-	signer, err := tss.GenerateThresholdSigner(
-		ctx,
-		keepAddress.Hex(),
-		memberID,
-		keepMembersIDs,
-		uint(len(keepMembersIDs)-1),
-		n.networkProvider,
-		n.tssParamsPool.get(),
-	)
-	if err != nil {
-		return nil, fmt.Errorf("failed to generate threshold signer: [%v]", err)
+	logger.Infof("Getting new key-generation pre-parameters from the pool")
+	preParamsBox := params.NewBox(n.tssParamsPool.get())
+	logger.Infof("Received new key-generation parameters from the pool")
+
+	attemptCounter := 0
+	for {
+		attemptCounter++
+
+		logger.Infof(
+			"signer generation for keep [%s]; attempt [%v]",
+			keepAddress.String(),
+			attemptCounter,
+		)
+
+		// If we are re-attempting the key generation, pre-parameters in the box
+		// could be destroyed because they were shared with other members.
+		// In this case, we need to re-generate them.
+		if preParamsBox.IsEmpty() {
+			logger.Infof("Getting new key-generation pre-parameters from the pool")
+			preParamsBox = params.NewBox(n.tssParamsPool.get())
+			logger.Infof("Received new key-generation parameters from the pool")
+		}
+
+		// Global timeout for generating a signer exceeded.
+		// We are giving up and leaving this function.
+		if ctx.Err() != nil {
+			return nil, fmt.Errorf("key generation timeout exceeded")
+		}
+
+		// Announce signer presence. Other members of the keep need to receive
+		// the public key of this members. This member, need to receive public
+		// keys of all other members. Up to this point, only addresses from
+		// signer selection protocol are known.
+		//
+		// If signer announcement fails, we retry from the beginning.
+		memberIDs, err := n.AnnounceSignerPresence(
+			ctx,
+			operatorPublicKey,
+			keepAddress,
+			members,
+		)
+		if err != nil {
+			logger.Errorf("failed to announce signer presence: [%v]", err)
```

Comment:
Since we're in a retry loop maybe we should log just a warning, not an error.

Reply:
I think it should be an error. The log level should be about a severity, not about the fact whether are going to retry or not. As an example, in `keep-core` we log on warning level when one of the nodes provided invalid message and this message was dropped and that node marked as misbehaving. This is not an error because the protocol can continue without that single misbehaving member with no problem. We warn that someone is misbehaving, though. In this case in `node.go`, the protocol cannot continue with all members announcing their presence so we should use error level. This is a kind of a situation that should potentially light the red bulb in the operator's room.